Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -892,63 +892,167 @@ extension SwiftLanguageServer {
     let providers = providersAndKinds.filter { wantedActionKinds?.contains($0.1) != false }
-    retrieveCodeActions(req, providers: providers.map { $0.provider }) { codeActions in
-      let capabilities = self.clientCapabilities.textDocument?.codeAction
-      let response = CodeActionRequestResponse(codeActions: codeActions,
-                                               clientCapabilities: capabilities)
-      req.reply(response)
+    retrieveCodeActions(req, providers: providers.map { $0.provider }) { result in
+      switch result {
+      case .success(let codeActions):
+        let capabilities = self.clientCapabilities.textDocument?.codeAction
+        let response = CodeActionRequestResponse(codeActions: codeActions,
+                                                 clientCapabilities: capabilities)
+        req.reply(response)
+      case .failure(let error):
+        req.reply(.failure(error))
+      }
     }
   }
 
   func retrieveCodeActions(_ req: Request<CodeActionRequest>, providers: [CodeActionProvider], completion: @escaping CodeActionProviderCompletion) {
     guard providers.isEmpty == false else {
-      completion([])
+      completion(.success([]))
       return
     }
     var codeActions = [CodeAction]()
     let dispatchGroup = DispatchGroup()
     (0..<providers.count).forEach { _ in dispatchGroup.enter() }
     dispatchGroup.notify(queue: queue) {
-      completion(codeActions)
+      completion(.success(codeActions))
     }
     for i in 0..<providers.count {
-      providers[i](req.params) { actions in
+      providers[i](req.params) { result in
+        defer { dispatchGroup.leave() }
+        guard case .success(let actions) = result else {
+          return
+        }
         self.queue.sync {
           codeActions += actions
         }
-        dispatchGroup.leave()
       }
     }
   }
 
+  func retrieveRefactorCodeActions(_ params: CodeActionRequest, completion: @escaping CodeActionProviderCompletion) {
+    guard let snapshot = documentManager.latestSnapshot(params.textDocument.url) else {
+      let message = "failed to find snapshot for url \(params.textDocument.url)"
+      log(message)
+      completion(.failure(.unknown(message)))
+      return
+    }
+
+    guard let offsetRange = snapshot.utf8OffsetRange(of: params.range) else {
+      let message = "failed to retrieve range from \(params.range)"
+      log(message)
+      completion(.failure(.unknown(message)))
+      return
+    }
+
+    let additionalCursorInfoParameters: ((SKRequestDictionary) -> Void) = { skreq in
+      skreq[self.keys.length] = offsetRange.count
+      skreq[self.keys.retrieve_refactor_actions] = 1
+    }
+
+    cursorInfo(
+      params.textDocument.url,
+      params.range.lowerBound,
+      customCursorOffset: offsetRange.lowerBound,
+      additionalParameters: additionalCursorInfoParameters)
+    { result in
+      guard let dict: CursorInfo = result.success ?? nil else {
```

Comment:
What is `?? nil` doing?

Reply:
The `result.success` here is a double optional -- I followed the same code style from hover/symbolinfo but I agree it's confusing.